libsplinter/src/auth/rest_api/identity/mod.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Tools for identifying clients and users
+
+mod error;
+#[cfg(feature = "oauth-github")]
+pub mod github;
+
+use std::str::FromStr;
+
+pub use error::{AuthorizationParseError, IdentityProviderError};
+
+/// A service that fetches identities from a backing provider
+pub trait IdentityProvider: Send {
+    /// Attempts to get the identity that corresponds to the given authorization
+    fn get_identity(&self, authorization: &Authorization) -> Result<String, IdentityProviderError>;
+
+    /// Clones implementation for `IdentityProvider`. The implementation of the `Clone` trait for
+    /// `Box<dyn IdentityProvider>` calls this method.
+    ///
+    /// # Example
+    ///
+    ///```ignore
+    ///  fn clone_box(&self) -> Box<dyn IdentityProvider> {
+    ///     Box::new(self.clone())
+    ///  }
+    ///```
+    fn clone_box(&self) -> Box<dyn IdentityProvider>;
+}
+
+impl Clone for Box<dyn IdentityProvider> {
+    fn clone(&self) -> Box<dyn IdentityProvider> {
+        self.clone_box()
+    }
+}
+
+/// The authorization that is passed to an `IdentityProvider`
+pub enum Authorization {
+    Bearer(BearerToken),
+}
+
+/// Parses an authorization string, which must be in the format "<scheme> <value>"
+impl FromStr for Authorization {
+    type Err = AuthorizationParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut parts = s.splitn(2, ':');
+        match (
+            parts.next().expect("cannot return empty iter"),
```

Comment:
this should be `ok_or_else(|| AuthorizationParseError...)`

Reply:
I.e. it implies an empty string

Reply:
Calling `splitn` on the empty string returns an iterator with a single item, the empty string. The iterator is never empty.